packages/next/src/client/components/router-reducer/reducers/create-prefetch-cache-key.ts

```diff
@@ -0,0 +1,26 @@
+import { addPathPrefix } from '../../../../shared/lib/router/utils/add-path-prefix'
+import { pathHasPrefix } from '../../../../shared/lib/router/utils/path-has-prefix'
+import { createHrefFromUrl } from '../create-href-from-url'
+
+/**
+ * Creates a cache key for the router prefetch cache
+ *
+ * @param url - The URL being navigated to
+ * @param nextUrl - an internal URL, primarily used for handling rewrites. Defaults to '/'.
+ * @return The generated prefetch cache key.
+ */
+export function createPrefetchCacheKey(url: URL, nextUrl: string | null) {
+  const pathnameFromUrl = createHrefFromUrl(
+    url,
+    // Ensures the hash is not part of the cache key as it does not impact the server fetch
+    false
+  )
+
+  // Route interception depends on `nextUrl` values which aren't a 1:1 mapping to a URL
+  // The cache key that we store needs to use `nextUrl` to properly distinguish cache entries
+  if (nextUrl && !pathHasPrefix(pathnameFromUrl, nextUrl)) {
+    return addPathPrefix(nextUrl, pathnameFromUrl)
+  }
```

Comment:
As this cache key doesn't reflect the real URL, maybe joining it via a method that wouldn't conflict would be more reliable?

For example, if the `pathname` is `/bar`, and the `nextUrl` is `/foo`, but there's a route for `/foo/bar`, then this would cause them to clash right? Perhaps if you did something like `%` so the cache key could be `/foo%/bar` which by itself isn't a valid path, but would allow the interception route to match.